pkg/services/sso/keycloak_service_builder.go

```diff
@@ -8,34 +8,103 @@ import (
 
 var _ KeycloakServiceBuilderSelector = &keycloakServiceBuilderSelector{}
 var _ KeycloakServiceBuilder = &keycloakServiceBuilder{}
+var _ KFMKeycloakServiceBuilderConfigurator = &keycloakBuilderConfigurator{}
+var _ OSDKeycloakServiceBuilderConfigurator = &osdBuilderConfigurator{}
+
+type KeycloakServiceBuilderSelector interface {
+	ForOSD() OSDKeycloakServiceBuilderConfigurator
+	ForKFM() KFMKeycloakServiceBuilderConfigurator
+}
+
+type KFMKeycloakServiceBuilderConfigurator interface {
+	WithConfiguration(config *keycloak.KeycloakConfig) KeycloakServiceBuilder
+}
+
+type OSDKeycloakServiceBuilderConfigurator interface {
+	WithConfiguration(config *keycloak.KeycloakConfig) OSDKeycloakServiceBuilder
+}
 
 type KeycloakServiceBuilder interface {
 	WithRealmConfig(realmConfig *keycloak.KeycloakRealmConfig) KeycloakServiceBuilder
 	Build() KeycloakService
 }
 
+type OSDKeycloakServiceBuilder interface {
+	WithRealmConfig(realmConfig *keycloak.KeycloakRealmConfig) OSDKeycloakServiceBuilder
+	Build() OSDKeycloakService
+}
+
+type keycloakServiceBuilderSelector struct {
+}
+
+func (s *keycloakServiceBuilderSelector) ForOSD() OSDKeycloakServiceBuilderConfigurator {
+	return &osdBuilderConfigurator{}
+}
+
+func (s *keycloakServiceBuilderSelector) ForKFM() KFMKeycloakServiceBuilderConfigurator {
+	return &keycloakBuilderConfigurator{}
+}
+
+type keycloakBuilderConfigurator struct{}
+type osdBuilderConfigurator keycloakBuilderConfigurator
+
+func (k *keycloakBuilderConfigurator) WithConfiguration(config *keycloak.KeycloakConfig) KeycloakServiceBuilder {
+	return &keycloakServiceBuilder{
+		config: config,
+	}
+}
+
+func (o *osdBuilderConfigurator) WithConfiguration(config *keycloak.KeycloakConfig) OSDKeycloakServiceBuilder {
+	return &osdKeycloackServiceBuilder{
+		config: config,
+	}
+}
+
 type keycloakServiceBuilder struct {
 	config      *keycloak.KeycloakConfig
 	realmConfig *keycloak.KeycloakRealmConfig
 }
 
+type osdKeycloackServiceBuilder keycloakServiceBuilder
+
 // Build returns an instance of KeycloakService ready to be used.
 // If a custom realm is configured (WithRealmConfig called), then always Keycloak provider is used
 // irrespective of the `builder.config.SelectSSOProvider` value
 func (builder *keycloakServiceBuilder) Build() KeycloakService {
+	return build(builder.config.SelectSSOProvider, builder.config, builder.realmConfig)
+}
+
+func (builder *keycloakServiceBuilder) WithRealmConfig(realmConfig *keycloak.KeycloakRealmConfig) KeycloakServiceBuilder {
+	builder.realmConfig = realmConfig
+	return builder
+}
+
+// Build returns an instance of KeycloakService ready to be used.
+// If a custom realm is configured (WithRealmConfig called), then always Keycloak provider is used
+// irrespective of the `builder.config.SelectSSOProvider` value
+func (builder *osdKeycloackServiceBuilder) Build() OSDKeycloakService {
+	return build(builder.config.SelectSSOProvider, builder.config, builder.realmConfig).(OSDKeycloakService)
+}
```

Comment:
The comment is not true for a keycloak service used for the configuration of the identity provider. 
For this service, we should always use the rhoas-kafka-sre role irrespective of the provider type i.e there is no need to pass `builder.config.SelectSSOProvider` or have a `WithRealmConfig(...)` that takes a realm in argument.